go-controller/pkg/util/node_annotations_unit_test.go

```diff
@@ -0,0 +1,298 @@
+	vlanid := uint(1024)
+	tests := []struct {
+		desc       string
+		inpL3GwCfg *L3GatewayConfig
+		expOutput  []byte
+	}{
+		{
+			desc:       "test empty config i.e gateway mode disabled by default",
+			inpL3GwCfg: &L3GatewayConfig{},
+			expOutput:  []byte(`{"mode":""}`),
+		},
+		{
+			desc: "test gateway mode set to local and verify that node-port-enable is set to false by default",
+			inpL3GwCfg: &L3GatewayConfig{
+				Mode: config.GatewayModeLocal,
+			},
+			expOutput: []byte(`{"mode":"local","node-port-enable":"false"}`),
+		},
+		{
+			desc: "test VLANID not nil",
+			inpL3GwCfg: &L3GatewayConfig{
+				Mode:   config.GatewayModeLocal,
+				VLANID: &vlanid,
+			},
+			expOutput: []byte(`{"mode":"local","node-port-enable":"false","vlan-id":"1024"}`),
+		},
+		{
+			desc: "test single IP address and single next hop path",
+			inpL3GwCfg: &L3GatewayConfig{
+				Mode:        config.GatewayModeLocal,
+				VLANID:      &vlanid,
+				IPAddresses: []*net.IPNet{ovntest.MustParseIPNet("192.168.1.10/24")},
+				NextHops:    []net.IP{ovntest.MustParseIP("192.168.1.1")},
+			},
+			expOutput: []byte(`{"mode":"local","ip-addresses":["192.168.1.10/24"],"ip-address":"192.168.1.10/24","next-hops":["192.168.1.1"],"next-hop":"192.168.1.1","node-port-enable":"false","vlan-id":"1024"}`),
+		},
+		{
+			desc: "test multiple IP address and multiple next hop paths",
+			inpL3GwCfg: &L3GatewayConfig{
+				Mode:        config.GatewayModeLocal,
+				VLANID:      &vlanid,
+				InterfaceID: "INTERFACE-ID",
+				MACAddress:  ovntest.MustParseMAC("11:22:33:44:55:66"),
+				IPAddresses: []*net.IPNet{
+					ovntest.MustParseIPNet("192.168.1.10/24"),
+					ovntest.MustParseIPNet("fd01::1234/64"),
+				},
+				NextHops: []net.IP{
+					ovntest.MustParseIP("192.168.1.1"),
+					ovntest.MustParseIP("fd01::1"),
+				},
+			},
+			expOutput: []byte(`{"mode":"local","interface-id":"INTERFACE-ID","mac-address":"11:22:33:44:55:66","ip-addresses":["192.168.1.10/24","fd01::1234/64"],"next-hops":["192.168.1.1","fd01::1"],"node-port-enable":"false","vlan-id":"1024"}`),
+		},
+	}
+	for i, tc := range tests {
+		t.Run(fmt.Sprintf("%d:%s", i, tc.desc), func(t *testing.T) {
+			res, e := tc.inpL3GwCfg.MarshalJSON()
+			t.Log(string(res), e)
+			assert.True(t, reflect.DeepEqual(res, tc.expOutput))
+		})
+	}
+}
+
+func TestL3GatewayConfig_UnmarshalJSON(t *testing.T) {
+	tests := []struct {
+		desc        string
+		expOut      L3GatewayConfig
+		inputParam  []byte
+		errExpected bool
+	}{
+		{
+			desc:        "error: test bad input causing json Unmarshal error",
+			errExpected: true,
+			inputParam:  []byte(`{`),
+		},
+		{
+			desc:       "success: test gateway mode disabled path",
+			inputParam: []byte(`{"mode":""}`),
+			expOut: L3GatewayConfig{
+				Mode:           "",
+				NodePortEnable: false,
+			},
+		},
+		{
+			desc:        "error: test unsupported gateway mode",
+			inputParam:  []byte(`{"mode":"blah"}`),
+			errExpected: true,
```

Comment:
It's generally better to assert more specific things about the error.

eg, if someone added a check that "mac-address" always has to be set, and then later broke the check for unsupported gateway mode, this test would still pass, because it would be returning a "mac address is unset" error, which is the wrong error, but would fool the test case.

in other places it's common to check for a substring match on the returned error

Reply:
Valid point. If an error message is originating inside of the function being unit tested, checking for substring match would be effective.